File: packages/server-php/src/MemoryChallengeStore.php
<?php

declare(strict_types=1);

namespace OpenPasskey\Server;

class MemoryChallengeStore implements ChallengeStore
{
    /** @var array<string, array{challenge: string, expiresAt: float}> */
    private array $entries = [];
    private int $writeCount = 0;

    public function store(string $key, string $challenge, float $timeoutSeconds): void
    {
        $this->entries[$key] = [
            'challenge' => $challenge,
            'expiresAt' => microtime(true) + $timeoutSeconds,
        ];

        $this->writeCount++;
        if ($this->writeCount % 100 === 0) {
            $this->cleanup();
        }
    }

    public function consume(string $key): string
    {
        if (!isset($this->entries[$key])) {
            throw new PasskeyError('challenge not found or expired');
        }

        $entry = $this->entries[$key];
        unset($this->entries[$key]);

        if (microtime(true) > $entry['expiresAt']) {
            throw new PasskeyError('challenge not found or expired');
        }

        return $entry['challenge'];
    }

    private function cleanup(): void
    {
        $now = microtime(true);
        foreach ($this->entries as $key => $entry) {
            if ($now > $entry['expiresAt']) {
                unset($this->entries[$key]);
            }
        }
    }
}

File: packages/server-rails/lib/open_passkey/stores/memory_challenge_store.rb
# frozen_string_literal: true

module OpenPasskey
  class MemoryChallengeStore
    include ChallengeStore

    def initialize
      @entries = {}
      @write_count = 0
    end

    def store(key, challenge, timeout_seconds)
      @entries[key] = {
        challenge: challenge,
        expires_at: Process.clock_gettime(Process::CLOCK_REALTIME) + timeout_seconds
      }

      @write_count += 1
      cleanup if (@write_count % 100).zero?
    end

    def consume(key)
      entry = @entries.delete(key)
      raise PasskeyError.new("challenge not found or expired") unless entry

      if Process.clock_gettime(Process::CLOCK_REALTIME) > entry[:expires_at]
        raise PasskeyError.new("challenge not found or expired")
      end

      entry[:challenge]
    end

    private

    def cleanup
      now = Process.clock_gettime(Process::CLOCK_REALTIME)
      @entries.delete_if { |_, entry| now > entry[:expires_at] }
    end
  end
end
